distributed/scheduler.py

```diff
@@ -2259,7 +2259,7 @@ def transition_processing_released(self, key: str, stimulus_id: str):
                 assert ts.state == "processing"
 
             w = _remove_from_processing(self, ts)
-            if w:
+            if w in self.workers:
```

Comment:
I'm a little concerned about this. It feels like another instance of https://github.com/dask/distributed/issues/6392.

In https://github.com/dask/distributed/pull/6614#discussion_r949755332 I have a driveby refactor fixing a possible stale WorkerState in `_remove_from_processing`.

If we go ahead and return the address even if we know it's stale, then `transition_processing_released` here has no ability to verify whether `w` is stale or not, because it just has an address, not a WorkerState instance. It's possible that a worker left, then reconnected with the same address and port. If that's the case, we'd then be sending a `free-keys` to the wrong worker instance.

I'm wondering if instead of the approach in this PR, we should go back to the old behavior and instead explicitly handle the case in `transition_processing_erred` where both `w` and `worker` are None (i.e. the KilledWorker case)?

Reply:
Two thoughts: 
1. IIUC, since we're not awaiting anywhere in https://github.com/dask/distributed/blob/c15a10e87ca5d03e62f0ad4f38adb63163522979/distributed/scheduler.py#L2261-L2262, it should never happen that https://github.com/dask/distributed/blob/c15a10e87ca5d03e62f0ad4f38adb63163522979/distributed/scheduler.py#L7337 is True while https://github.com/dask/distributed/blob/c15a10e87ca5d03e62f0ad4f38adb63163522979/distributed/scheduler.py#L2262 is False. Thus, this piece of code should behave the same way as before.
2. Instead of handling the `None` case, should we maybe return the entire `WorkerState` object to ensure that users of `_remove_from_processing` can check for stale worker states as you described? From your description, it sounds like returning a worker adrees might generally not be all too helpful.

Reply:
Note: I wrote the previous response without looking at the linked comment. That drive-by fix makes sense to me. To solve the issue at hand, I'd return whatever https://github.com/dask/distributed/blob/c15a10e87ca5d03e62f0ad4f38adb63163522979/distributed/scheduler.py#L7333-L7334 spits out. This should allow downstream users to differentiate between stale and current worker states.

Reply:
I guess I don't like downstream users having to differentiate themselves (more points to mess that up). Returning None felt pretty sensible to me in that case. In the rare case where callers need to know the worker we thought it was processing on, even if that worker might be stale—I believe `transition_processing_erred` is the only case where this applies?—why not retrieve that explicitly first?